File: objc/objcfind/Sources/objcfind/FileResultSorter.m
//
//  FileResultSorter.m
//  objcfind
//
//  Created by Cary Clark on 7/1/25.
//

#import <Foundation/Foundation.h>
#import "FileResultSorter.h"
#import "FileUtil.h"

@implementation FileResultSorter

- (instancetype) initWithSettings:(FindSettings*)settings {
    self = [super init];
    if (self) {
        self.settings = settings;
    }
    return self;
}

- (NSComparisonResult (^)(FileResult*, FileResult*)) getFileResultComparator {
    if (self.settings.sortDescending) {
        if (self.settings.sortBy == SortByFileName) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr2 compareByName:fr1 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else if (self.settings.sortBy == SortByFileSize) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr2 compareBySize:fr1 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else if (self.settings.sortBy == SortByFileType) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr2 compareByType:fr1 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else if (self.settings.sortBy == SortByLastMod) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr2 compareByLastMod:fr1 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr2 compareByPath:fr1 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        }
    } else {
        if (self.settings.sortBy == SortByFileName) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr1 compareByName:fr2 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else if (self.settings.sortBy == SortByFileSize) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr1 compareBySize:fr2 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else if (self.settings.sortBy == SortByFileType) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr1 compareByType:fr2 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else if (self.settings.sortBy == SortByLastMod) {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr1 compareByLastMod:fr2 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        } else {
            return ^NSComparisonResult(FileResult *fr1, FileResult *fr2) {
                return [fr1 compareByPath:fr2 caseInsensitive:self.settings.sortCaseInsensitive];
            };
        }
    }
}

- (NSArray<FileResult*>*) sort:(NSArray<FileResult*>*)fileResults {
    return [fileResults sortedArrayUsingComparator:[self getFileResultComparator]];
}

@end

File: java/javafind/lib/src/main/java/javafind/FileResultSorter.java
package javafind;

import java.util.Comparator;
import java.util.List;

public class FileResultSorter {
    private final FindSettings settings;

    public FileResultSorter(final FindSettings settings) {
        this.settings = settings;
    }

    public FindSettings getSettings() {
        return settings;
    }

    public final Comparator<FileResult> getFileResultComparator() {
        var sortBy = settings.getSortBy() == null ? SortBy.FILEPATH : settings.getSortBy();
        if (settings.getSortDescending()) {
            return switch (sortBy) {
                case FILENAME -> (FileResult fr1, FileResult fr2) -> fr2.compareByName(fr1, settings.getSortCaseInsensitive());
                case FILESIZE -> (FileResult fr1, FileResult fr2) -> fr2.compareBySize(fr1, settings.getSortCaseInsensitive());
                case FILETYPE -> (FileResult fr1, FileResult fr2) -> fr2.compareByType(fr1, settings.getSortCaseInsensitive());
                case LASTMOD -> (FileResult fr1, FileResult fr2) -> fr2.compareByLastMod(fr1, settings.getSortCaseInsensitive());
                default -> (FileResult fr1, FileResult fr2) -> fr2.compareByPath(fr1, settings.getSortCaseInsensitive());
            };
        }
        return switch (sortBy) {
            case FILENAME -> (FileResult fr1, FileResult fr2) -> fr1.compareByName(fr2, settings.getSortCaseInsensitive());
            case FILESIZE -> (FileResult fr1, FileResult fr2) -> fr1.compareBySize(fr2, settings.getSortCaseInsensitive());
            case FILETYPE -> (FileResult fr1, FileResult fr2) -> fr1.compareByType(fr2, settings.getSortCaseInsensitive());
            case LASTMOD -> (FileResult fr1, FileResult fr2) -> fr1.compareByLastMod(fr2, settings.getSortCaseInsensitive());
            default -> (FileResult fr1, FileResult fr2) -> fr1.compareByPath(fr2, settings.getSortCaseInsensitive());
        };
    }

    public final void sort(List<FileResult> fileResults) {
        if (fileResults.isEmpty()) {
            return;
        }
        var fileResultComparator = getFileResultComparator();
        fileResults.sort(fileResultComparator);
    }
}
